nova/scheduler/filters/external_scheduler_filter.py

```diff
@@ -0,0 +1,78 @@
+# Copyright 2024 SAP SE or an SAP affiliate company.
+#
+#    Licensed under the Apache License, Version 2.0 (the "License"); you may
+#    not use this file except in compliance with the License. You may obtain
+#    a copy of the License at
+#
+#         http://www.apache.org/licenses/LICENSE-2.0
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+#    WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+#    License for the specific language governing permissions and limitations
+#    under the License.
+
+
+import requests
+
+import nova.conf
+from nova.scheduler import filters
+from oslo_log import log as logging
+
+LOG = logging.getLogger(__name__)
+CONF = nova.conf.CONF
+
+
+class ExternalSchedulerFilter(filters.BaseHostFilter):
+    """Filter using an out of tree scheduler via a REST API."""
+
+    RUN_ON_REBUILD = False
+
+    def __init__(self):
+        super().__init__()
+        self.api_url = CONF.filter_scheduler.external_scheduler_api_url
+
+    def filter_all(self, filter_obj_list, spec_obj):
+        """Yield objects that pass the filter."""
+        if not self.api_url:
+            LOG.debug(
+                "external_scheduler_api_url not configured. skipping filter"
+            )
+            return filter_obj_list
+
+        hosts_data = []
+        for obj in filter_obj_list:
+            hosts_data.append({
+                "name": obj.host,
+                "status": obj.status,
+                "vcpus_total": obj.vcpus_total,
+                "vcpus_used": obj.vcpus_used,
+                "memory_mb": obj.memory_mb,
+                "memory_mb_used": obj.memory_mb_used
+            })
+
+        spec_data = {
+            "vcpus": spec_obj.vcpus,
+            "memory_mb": spec_obj.memory_mb,
+        }
+
+        json_data = {
+            "hosts": hosts_data,
+            "request_spec": spec_data
+        }
+
+        try:
+            response = requests.post(
+                self.api_url, json=json_data, timeout=10
+            )
+            response.raise_for_status()
+            response_json = response.json()
+            valid_hosts = response_json.get('valid_hosts', [])
+            for obj in filter_obj_list:
+                if obj.host in valid_hosts:
+                    yield obj
+
+        except requests.RequestException as e:
+            # Log an error if the request fails
+            LOG.error("Failed to query the external API: %s", e)
+            return filter_obj_list
```

Comment:
I think this does nothing. I tried with this code:
```python
def a():
  try:
      for i in range(7):
        if i == 3:
          raise Exception
        yield i
  except Exception:
    return list(range(7))

breakpoint()
```
```python
$ python3 /tmp/bla.py 
--Return--
> /tmp/bla.py(10)<module>()->None
-> breakpoint()
(Pdb) p a()
<generator object a at 0x7e701c04d4d0>
(Pdb) p list(a())
[0, 1, 2]
```
Since a function containing `yield`, not matter if it's called or not, is a Generator, it doesn't have a proper return-value - afaik.

Any return is broken, also the first one allowing to skip the filter:
```python
def a():
  return ["early"]
  try:
      for i in range(7):
        if i == 3:
          raise Exception
        yield i
  except Exception:
    return list(range(7))

breakpoint()
```
```python
$ python3 /tmp/bla.py 
--Return--
> /tmp/bla.py(11)<module>()->None
-> breakpoint()
(Pdb) p list(a())
[]
(Pdb) 
```

Am I holding it wrong? Shouldn't the test for "the filter returns everything if it's disabled" have found this?

Reply:
Since you have all the hosts in a list already and are not reading from the request anymore, you might as well return a list instead of yielding, i.e. `return [obj for obj in filter_obj_list if obj.host in valid_hosts]`